### PR TITLE
cwl: support tool edits

### LINKED ISSUE #295: allow tool edits
**Is your feature request related to a problem? Please describe.**
In a typical exploratory phase of a project, it is quite common to rename/add/remove variables or parameters. We should allow users to do so for the CWL tools that are created as a result of `renku run` commands. 

**Describe the solution you'd like**

The full implementation will allow something like this:

```
renku run echo "woop" > out
renku tool list
id |  command  |  parameters | outputs
--------------------------------------
1  |  echo     |     woop    |  out

renku tool edit 1 --output 1,output_file

renku update
cat output_file
woop
```

A minimal implementation would allow me to edit the CWL file manually and allow for renku commands to work with the updated CWL description.

### PR DESCRIPTION
**Pull request by jirikuncar**
# Description

Simplifies status detection and adds test for usage of modified tool.

Fixes #295 

## Type of change

Please select relevant options.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] I have read and understood the **CONTRIBUTING** document.
- [x] I have performed a self-review of my own code.
- [x] I have signed and sent the [CLA document]
      (https://github.com/SwissDataScienceCenter/documentation/wiki/files/SDSC_Contributor_License_Agreement_v1.0.pdf).
- [x] I have added tests to cover my changes.
- [ ] I have **NOT** removed any existing tests.
- [ ] I have updated the documentation accordingly.
- [ ] I have described *all* changes in the **CHANGELOG.rst**.
- [x] I have run ``./run-tests.sh`` locally.

### CODE REVIEW

File: renku/cli/_graph.py [hunk 1]
```diff
@@ -67,62 +67,55 @@ def __attrs_post_init__(self):
...
     def default_nodes(self):
         """Build node index."""
-        _nodes = OrderedDict()
-        _need_update = self._need_update = {}
-        _latest_commits = self._latest_commits
+        nodes = []
+        for commit in reversed(self._sorted_commits):
+            try:
+                activity = self.commits[commit]
+                nodes.extend(reversed(list(activity.nodes)))
```
Comment (rokroskar) [inline renku/cli/_graph.py]: don't need `list`

File: renku/cli/_graph.py [hunk 2]
```diff
@@ -67,62 +67,55 @@ def __attrs_post_init__(self):
...
+        need_update_ = []
+        for parent in node.parents:
+            parent_updates = self.need_update(parent)
+            if parent_updates:
+                need_update_.extend(parent_updates)
+
+        return self._need_update.setdefault(node._id, need_update_)
+
+    def latest(self, node):
+        """Check if the node is in the latest version."""
```
Comment (rokroskar) [inline renku/cli/_graph.py]: The docstring implies that the return value should be a boolean, but instead it's the latest commit for this node.

File: renku/api/repository.py [hunk 3]
```diff
@@ -139,16 +139,23 @@ def project(self):
...
 
         if url:
             remote_url = 'https://' + url.hostname
-            if url.pathname:
-                remote_url += '/' + url.pathname
 
-            if remote_url.endswith('.git'):
-                remote_url = remote_url[:-4]
+            if url.owner:
+                remote_url += '/' + url.owner
```
[the remark this answers is not in the document]
Comment (jirikuncar) [reply renku/api/repository.py]: there won't be any - see the regex

File: renku/cli/_graph.py (hunk 1 above)
Comment (jirikuncar) [reply renku/cli/_graph.py]: it does need the `list` since `activity.nodes` might be a generator.

Comment (rokroskar) [reply renku/cli/_graph.py]: I know, but `extend` will take care of it, no?

Comment (jirikuncar) [reply renku/cli/_graph.py]: but not `reversed`

Comment (rokroskar) [reply renku/cli/_graph.py]: `reversed` can take a generator:

```
In [1]: reversed(range(10))
Out[1]: <range_iterator at 0x10acd79c0>

In [2]: x = []

In [3]: x.extend(reversed(range(10)))

In [4]: x
Out[4]: [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
```

Comment (rokroskar) [reply renku/cli/_graph.py]: I stand corrected, `range` is a special case ;)

**Pull request merged**